File: src/main/java/edu/practice/datastructures/linkedlist/DoublyLinkedList.java
package edu.practice.datastructures.linkedlist;

import java.util.NoSuchElementException;
import java.util.Objects;

public class DoublyLinkedList<E> {

    private static class Node<E> {
        E data;
        Node<E> prev;
        Node<E> next;

        Node(Node<E> prev, E data, Node<E> next) {
            this.prev = prev;
            this.data = data;
            this.next = next;
        }
    }

    private int size;
    private Node<E> first;
    private Node<E> last;

    public DoublyLinkedList() {// Initialize empty Linked List
    }

    public int size() { return size; }

    public boolean isEmpty() {
        return (size == 0);
    }

    public void addFirst(E item) {
        first = new Node<>(null, item, first);

        if (last == null) {
            last = first;
        }

        size++;
    }

    public void addLast(E item) {
        // NOTE: I had difficulty in modifying SLL's addLast method to convert to DLL.
        // This was done after 2 failed tests
        // Remember to go through this once again.
        
        final Node<E> newNode = new Node<>(last, item, null);

        if (last == null) {
            first = newNode;
        } else {
            last.next = newNode;
        }

        last = newNode;

        size++;
    }

    public E getFirst() {
        if (first == null) {
            throw new NoSuchElementException();
        }

        return first.data;
    }

    public E getLast() {
        if (last == null) {
            throw new NoSuchElementException();
        }

        return last.data;
    }

    public E removeFirst() {
        final E item = first.data;
        first = first.next;

        // First is now pointing to previously 2nd element of list, so we check if there was no other element
        if (first == null) {
            last = null;
        }

        size--;

        return item;
    }

    public E removeLast() {
        final E item = last.data;

        last = last.prev;
        if (last == null) { // That is, if there was only one node in the list
            first = null;
        } else {
            last.next = null;
        }

        size--;

        return item;
    }

    private boolean existsInList(E item, Node<E> firstNode) {
        if (firstNode == null) {
            return false;
        }

        if (Objects.equals(item, firstNode.data)) {
            return true;
        }

        return existsInList(item, firstNode.next);
    }

    public boolean contains(E item) {
        return existsInList(item, first);
    }
}

File: src/main/kotlin/datastructures/linkedlist/DoublyLinkedList.kt
package edu.practice.datastructures.linkedlist

class DoublyLinkedList<E>: LinkedList<E> {

    private data class Node<E>(val data: E, var prev: Node<E>? = null, var next: Node<E>? = null)

    private var size = 0

    private var first: Node<E>? = null
    private var last: Node<E>? = null

    override fun size(): Int = size

    override fun isEmpty(): Boolean = (size == 0)

    override fun getFirst(): E {
        return first?.data ?: throw NoSuchElementException("Linked List is Empty")
    }

    override fun getLast(): E {
        return last?.data ?: throw NoSuchElementException("Linked List is Empty")
    }

    override fun addFirst(item: E) {
        val newNode = Node(item)
        newNode.next = first
        first = newNode

        if (last == null) { // Empty DLL condition
            last = newNode
        }

        size++
    }

    override fun addLast(item: E) {
        val newNode = Node(item, last)

        if (last == null) { // Empty DLL condition
            first = newNode
        } else {
            last?.next = newNode
        }

        last = newNode
        size++
    }

    override fun removeFirst(): E {
        val item = getFirst()

        first = first?.next
        first?.prev = null

        if (first == null) {
            last = null
        }

        size--
        return item
    }

    override fun removeLast(): E {
        val item = getLast()

        last = last?.prev
        last?.next = null

        if (last == null) {
            first = null
        }

        size--
        return item
    }

    private fun existsInList(item: E, firstNode: Node<E>? = first): Boolean {
        return firstNode?.let {
            it.data == item || existsInList(item, it.next)
        } ?: false
    }

    override fun contains(item: E): Boolean {
        return existsInList(item)
    }
}